src/transaction.rs

```diff
@@ -10,78 +10,129 @@ pub const SIGNED_DATA_OFFSET: usize = size_of::<Signature>();
-    /// * `transaction_keys` - The keys for the transaction.  These are the program state
+    /// * `account_keys` - The keys for the transaction.  These are the program state
     ///    instances or token recipient keys.
-    /// * `userdata` - The input data that the program will execute with
     /// * `last_id` - The PoH hash.
     /// * `fee` - The transaction fee.
+    /// * `program_keys` - The keys that identify programs used in the `instruction` vector.
+    /// * `instructions` - The programs and their arguments that the transaction will execute atomically
     pub fn new(
         from_keypair: &Keypair,
-        transaction_keys: &[Pubkey],
-        program_id: Pubkey,
-        userdata: Vec<u8>,
+        keys: &[Pubkey],
         last_id: Hash,
         fee: i64,
+        program_keys: Vec<Pubkey>,
+        instructions: Vec<Instruction>,
     ) -> Self {
         let from = from_keypair.pubkey();
-        let mut keys = vec![from];
-        keys.extend_from_slice(transaction_keys);
+        let mut account_keys = vec![from];
+        account_keys.extend_from_slice(keys);
         let mut tx = Transaction {
             signature: Signature::default(),
-            keys,
-            program_id,
+            account_keys,
             last_id,
             fee,
-            userdata,
+            program_keys,
+            instructions,
         };
         tx.sign(from_keypair);
         tx
     }
+    pub fn new_one_program(
+        from_keypair: &Keypair,
+        transaction_keys: &[Pubkey],
+        program_id: Pubkey,
+        userdata: Vec<u8>,
+        last_id: Hash,
+        fee: i64,
+    ) -> Self {
+        let program_keys = vec![program_id];
+        let instructions = vec![Instruction {
+            program_id: 0,
+            userdata,
+            accounts: (0..(transaction_keys.len() as u8 + 1))
+                .into_iter()
+                .collect(),
+        }];
+        Self::new(
+            from_keypair,
+            transaction_keys,
+            last_id,
+            fee,
+            program_keys,
+            instructions,
+        )
+    }
 
+    pub fn userdata(&self, pix: usize) -> &[u8] {
+        &self.instructions[pix].userdata
+    }
+    pub fn key(&self, pix: usize, kix: usize) -> Option<&Pubkey> {
+        self.instructions
+            .get(pix)
+            .and_then(|p| p.accounts.get(kix))
+            .and_then(|ai| self.account_keys.get(*ai as usize))
+    }
+    pub fn program_id(&self, pix: usize) -> &Pubkey {
+        &self.program_keys[self.instructions[pix].program_id as usize]
+    }
     /// Get the transaction data to sign.
     pub fn get_sign_data(&self) -> Vec<u8> {
-        let mut data = serialize(&(&self.keys)).expect("serialize keys");
-
-        let program_id = serialize(&(&self.program_id)).expect("serialize program_id");
-        data.extend_from_slice(&program_id);
+        let mut data = serialize(&(&self.account_keys)).expect("serialize account_keys");
```

Comment:
Did you test this? The double-reference will work since `serialize` will accept any type, but is it useful? How does it affect the serialization format? If I remove it, what breaks?